hphp/runtime/ext/std/ext_std_process.cpp

```diff
@@ -201,7 +203,13 @@ struct ShellExecContext final {
 #endif
     }
     if (m_sig_handler) {
```

Comment:
These handlers are of the type signalhandler_t, which is a function pointer.

    typedef void (*__sighandler_t) (int);

The interesting thing is that they can be attached as a plain signal or as a `sigaction`. In the case of `light-process`, it's attached as a `sigaction`. `sigaction`s take 3 parameters and not one. Here's the `sigaction` structure that holds the details about an attached signal or `sigaction`

    struct sigaction {
      void     (*sa_handler)(int);
      void     (*sa_sigaction)(int, siginfo_t *, void *);
      sigset_t   sa_mask;
      int        sa_flags;
      void     (*sa_restorer)(void);
    };

Some platforms will union the `sa_handler` and `sa_action` above since they're mutually exclusive. So when the `ShellExecContext` ctor temporarily attaches the generic handler with `signal(SIGCHLD, SIG_DFL)`, the old handler that is returned is a function pointer to `LightProcess::SigChldHandler()`, which is a `sigaction` taking three parameters. This means that we can't restore this handler using `signal(SIGCHLD, m_sig_handler)`, because doing so would then register that function pointer as a simple `signal` and not a `sigaction`. So what's happening here is that `LightProcess::SigChldHandler()` is being called as if it were a simple `signal` with only one parameter instead of three, leading to garbage in the other two argument registers being used.

Sorry if you already got the above, but I'm finally answering your question now. :-) We can use the old handler value here, but the way to use it is to do something like the following:

    struct sigaction sa = {};
    sa.sa_sigaction = m_sig_handler;
    sa.sa_flags = SA_SIGINFO | SA_NOCLDSTOP;
    if (sigaction(SIGCHLD, &sa, nullptr) != 0) {
      Logger::Error("Couldn't install SIGCHLD handler");
      abort();
    }

This is basically a duplicate of the code in `light-process` that attached the `sigaction` originally. We just removed duplication by putting it in a new member method, `void LightProcess::AttachHandler(void)`. It's effectively the same thing as using the old handler value because `m_sig_handler` equals `&LightProcess::SigChldHandler`.